src/we-blog-remove.1
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.32
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "WE-BLOG-REMOVE 1"
.TH WE-BLOG-REMOVE 1 "2012-07-10" "Version 0.9" "We-Blog Documentation"
.SH "NAME"
we\-blog\-remove  \- removes a post or page from the We\-Blog repository
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBwe-blog-remove\fR [\fB\-fipqPV\fR] [\fB\-b\fR \fIdirectory\fR] \fIid\fR...
.PP
\&\fBwe-blog-remove\fR \fB\-h\fR|\fB\-v\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBwe-blog-remove\fR removes a blog post or a page with the specified \fIid\fR from
the We-Blog repository.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-b\fR \fIdirectory\fR, \fB\-\-blogdir\fR \fIdirectory\fR" 4
.IX Item "-b directory, --blogdir directory"
Allows you to specify a \fIdirectory\fR in which the We-Blog repository
is placed. The default option is a current working directory.
.IP "\fB\-p\fR, \fB\-\-page\fR, \fB\-\-pages\fR" 4
.IX Item "-p, --page, --pages"
Tells \fBwe-blog-remove\fR to remove a page or pages.
.IP "\fB\-P\fR, \fB\-\-post\fR, \fB\-\-posts\fR" 4
.IX Item "-P, --post, --posts"
Tells \fBwe-blog-remove\fR to remove a blog post or blog posts. This is the
default option.
.IP "\fB\-f\fR, \fB\-\-force\fR" 4
.IX Item "-f, --force"
Disables requiring manual confirmation of each blog post or page removal.
This is the default option.
.IP "\fB\-i\fR, \fB\-\-interactive\fR" 4
.IX Item "-i, --interactive"
Enables requiring manual confirmation of each blog post or page removal.
.IP "\fB\-q\fR, \fB\-\-quiet\fR" 4
.IX Item "-q, --quiet"
Disables displaying of unnecessary messages.
.IP "\fB\-V\fR, \fB\-\-verbose\fR" 4
.IX Item "-V, --verbose"
Enables displaying of all messages. This is the default option.
.IP "\fB\-h\fR, \fB\-\-help\fR" 4
.IX Item "-h, --help"
Displays usage information and exits.
.IP "\fB\-v\fR, \fB\-\-version\fR" 4
.IX Item "-v, --version"
Displays version information and exits.
.SH "EXAMPLE USAGE"
.IX Header "EXAMPLE USAGE"
Remove a blog post:
.PP
.Vb 2
\&        $ we-blog-remove 10
\&        Successfully removed the post with ID 10.
.Ve
.PP
Remove a page:
.PP
.Vb 2
\&        $ we-blog-remove -p 4
\&        Successfully removed the page with ID 4.
.Ve
.PP
Remove multiple blog posts:
.PP
.Vb 2
\&        $ we-blog-remove 10 4 6
\&        Successfully removed the post with ID 10, 4 and 6.
.Ve
.PP
Remove multiple blog posts safely:
.PP
.Vb 5
\&        $ we-blog-remove -i 10 4 6
\&        Remove the post with ID 10 titled `Debian and Fedora Packages'? y
\&        Remove the post with ID 4 titled `We-Blog 0.8.0 RC2'? y
\&        Remove the post with ID 6 titled `We-Blog 0.8.1'? y
\&        Successfully removed the post with ID 10, 4 and 6.
.Ve
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fBwe-blog-config\fR(1), \fBwe-blog-add\fR(1), \fBwe-blog-list\fR(1)
.SH "BUGS"
.IX Header "BUGS"
To report a bug or to send a patch, please, add a new issue to the bug
tracker at <http://code.google.com/p/we\-blog/issues/>, or visit the
discussion group at <https://groups.google.com/d/forum/tonk\-we\-blog>.
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright (c) 2008\-2011 Jaromir Hradilek / 2011\-2012 Ton Kersten
.PP
This program is free software; see the source for copying conditions. It is
distributed in the hope that it will be useful, but \s-1WITHOUT\s0 \s-1ANY\s0 \s-1WARRANTY\s0;
without even the implied warranty of \s-1MERCHANTABILITY\s0 or \s-1FITNESS\s0 \s-1FOR\s0 A
\&\s-1PARTICULAR\s0 \s-1PURPOSE\s0.


ChangeLog
2012-07-10  Ton Kersten <we-blog@tonkersten.com>
  * Declared 0.8 stable
  * Switched to branch 0.9 for further development

2012-07-09  Ton Kersten <we-blog@tonkersten.com>
  * Changed a lot of things, but they all boil down to
    bugfixes and code reshuffeling

2011-10-07  Ton Kersten <we-blog@tonkersten.com>
  * Added 'alt' text for pages to show on mouse-over.

2011-09-21  Ton Kersten <we-blog@tonkersten.com>
  * And switched to branch 0.8 for further development

2011-09-21  Ton Kersten <we-blog@tonkersten.com>
  * Increased we-blog to version 0.7
  * Added 'smiley' converter
  * Published the project on GitHub

2011-08-15  Ton Kersten <we-blog@tonkersten.com>
  * Forked BlazeBlogger 1.2.0 to we-blog 0.5
    This because development of Blaze seems to be stalled

2011-02-18  Jaromir Hradilek <jhradilek@gmail.com>

  * INSTALL, Makefile, README, src/we-blog-add.pl, src/we-blog-config.pl,
  src/we-blog-edit.pl, src/we-blog-init.pl, src/we-blog-list.pl, src/we-blog-log.pl,
  src/we-blog-make.pl, src/we-blog-remove.pl, unix/we-blog.sh, unix/man/man1/we-blog.1,
  windows/we-blog.bat: Increased the version number.

  * unix/man/man1/we-blog.1: Updated the manual page for the command wrapper.

# $Id: ChangeLog 4 2012-07-10 09:34:11 tonk $


src/we-blog-remove.pl
#!/usr/bin/env perl
# vi: set sw=4 ts=4 ai:
# $Id: we-blog-remove.pl 2 2011-09-21 15:34:50 tonk $

# we-blog-remove - removes a post or page from the We-Blog repository
# Copyright (c) 2011-2012 Ton Kersten
# Copyright (c) 2008-2011 Jaromir Hradilek

# This program is  free software:  you can redistribute it and/or modify it
# under  the terms  of the  GNU General Public License  as published by the
# Free Software Foundation, version 3 of the License.
#
# This program  is  distributed  in the hope  that it will  be useful,  but
# WITHOUT  ANY WARRANTY;  without  even the implied  warranty of MERCHANTA-
# BILITY  or  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
# License for more details.
#
# You should have received a copy of the  GNU General Public License  along
# with this program. If not, see <http://www.gnu.org/licenses/>.

use strict;
use warnings;
use File::Basename;
use File::Spec::Functions;
use Getopt::Long;

# Set the library path and use our own module
use lib dirname($0);
use we;

# Command-line options:
my  $type    = 'post';                              # Type: post or page.
my  $removed = '';                                  # List of removed IDs.

# Display usage information:
sub display_help {
	# Display the usage:
	print << "END_HELP";
Usage: $NAME [-fipqPV] [-b DIRECTORY] ID...
			 $NAME -h|-v

	-b, --blogdir directory     specify a directory in which the We-Blog
	                            repository is placed
	-p, --page                  remove a page or pages
	-P, --post                  remove a blog post or blog posts
	-i, --interactive           require manual confirmation of each removal
	-f, --force                 do not require manual confirmation
	-q, --quiet                 do not display unnecessary messages
	-V, --verbose               display all messages
	-h, --help                  display this help and exit
	-v, --version               display version information and exit
END_HELP

	# Return success:
	return 1;
}

# Remove a record or records from the repository:
sub remove_records {
	my $type = shift || 'post';
	my $ids  = shift || die 'Missing argument';

	# Initialize required variables:
	my @list = ();

	# Process each record:
	foreach my $id (@$ids) {
		# Prepare the file names:
		my $head = catfile($blogdir, $weblog, "${type}s", 'head', $id);
		my $body = catfile($blogdir, $weblog, "${type}s", 'body', $id);
		my $raw  = catfile($blogdir, $weblog, "${type}s", 'raw', $id);

		# Enter the interactive mode if requested:
		if ($prompt) {
			# Parse header data:
			my $data = read_ini($head);

			# Check whether the ID exists:
			unless ($data) {
				# Display the appropriate warning:
				display_warning("Unable to read the $type with ID $id.");

				# Move on to the next ID:
				next;
			}

			# Display the prompt:
			print "Remove $type with ID $id titled `" .
						($data->{header}->{title} || '') . "'? ";

			# Skip removal unless confirmed:
			next unless (readline(*STDIN) =~ /^(y|yes)$/i);
		}

		# Try to remove the record header:
		if (unlink $head) {
			# Remove the remaining record data:
			unlink($body, $raw);

			# Add the record to the list of successfully removed IDs:
			push(@list, $id);
		}
		else {
			# Report failure:
			display_warning("Unable to remove the $type with ID $id.");
		}
	}

	# Return the list of removed IDs:
	return @list;
}

# Set up the option parser:
Getopt::Long::Configure('no_auto_abbrev', 'no_ignore_case', 'bundling');

# Process command line options:
GetOptions(
	'help|h'        => sub { display_help();    exit 0; },
	'version|v'     => sub { display_version(); exit 0; },
	'page|pages|p'  => sub { $type    = 'page'; },
	'post|posts|P'  => sub { $type    = 'post'; },
	'force|f'       => sub { $prompt  = 0;      },
	'interactive|i' => sub { $prompt  = 1;      },
	'quiet|q'       => sub { $verbose = 0;      },
	'verbose|V'     => sub { $verbose = 1;      },
	'blogdir|b=s'   => sub { $blogdir = $_[1];  },
);

# Check missing options:
exit_with_error("Wrong number of options.", 22) if (scalar(@ARGV) < 1);

# Check whether the repository is present, no matter how naive this method
# actually is:
exit_with_error("Not a We-Blog repository! Try `we-blog-init' first.",1)
	unless (-d catdir($blogdir, ));

# Remove the records from the repository:
my @list = remove_records($type, \@ARGV);

# End here unless at least one record was actually removed:
unless (@list) {
	# Report abortion:
	print "Aborted.\n" if $verbose;

	# Return failure or success:
	exit (($prompt) ? 0 : 13);
}

# Prepare the list of successfully removed IDs:
$removed =  join(', ', sort(@list));
$removed =~ s/, ([^,]+)$/ and $1/;

# Log the event:
add_to_log("Removed the $type with ID $removed.")
	or display_warning("Unable to log the event.");

# Report success:
print "Successfully removed the $type with ID $removed.\n" if $verbose;

# Return success:
exit 0;

__END__

=head1 NAME

we-blog-remove  - removes a post or page from the We-Blog repository

=head1 SYNOPSIS

B<we-blog-remove> [B<-fipqPV>] [B<-b> I<directory>] I<id>...

B<we-blog-remove> B<-h>|B<-v>

=head1 DESCRIPTION

B<we-blog-remove> removes a blog post or a page with the specified I<id> from
the We-Blog repository.

=head1 OPTIONS

=over

=item B<-b> I<directory>, B<--blogdir> I<directory>

Allows you to specify a I<directory> in which the We-Blog repository
is placed. The default option is a current working directory.

=item B<-p>, B<--page>, B<--pages>

Tells B<we-blog-remove> to remove a page or pages.

=item B<-P>, B<--post>, B<--posts>

Tells B<we-blog-remove> to remove a blog post or blog posts. This is the
default option.

=item B<-f>, B<--force>

Disables requiring manual confirmation of each blog post or page removal.
This is the default option.

=item B<-i>, B<--interactive>

Enables requiring manual confirmation of each blog post or page removal.

=item B<-q>, B<--quiet>

Disables displaying of unnecessary messages.

=item B<-V>, B<--verbose>

Enables displaying of all messages. This is the default option.

=item B<-h>, B<--help>

Displays usage information and exits.

=item B<-v>, B<--version>

Displays version information and exits.

=back

=head1 EXAMPLE USAGE

Remove a blog post:

	$ we-blog-remove 10
	Successfully removed the post with ID 10.

Remove a page:

	$ we-blog-remove -p 4
	Successfully removed the page with ID 4.

Remove multiple blog posts:

	$ we-blog-remove 10 4 6
	Successfully removed the post with ID 10, 4 and 6.

Remove multiple blog posts safely:

	$ we-blog-remove -i 10 4 6
	Remove the post with ID 10 titled `Debian and Fedora Packages'? y
	Remove the post with ID 4 titled `We-Blog 0.8.0 RC2'? y
	Remove the post with ID 6 titled `We-Blog 0.8.1'? y
	Successfully removed the post with ID 10, 4 and 6.

=head1 SEE ALSO

B<we-blog-config>(1), B<we-blog-add>(1), B<we-blog-list>(1)

=head1 BUGS

To report a bug or to send a patch, please, add a new issue to the bug
tracker at <http://code.google.com/p/we-blog/issues/>, or visit the
discussion group at <http://groups.google.com/group/we-blog/>.

=head1 COPYRIGHT

Copyright (c) 2008-2011 Jaromir Hradilek / 2011-2012 Ton Kersten

This program is free software; see the source for copying conditions. It is
distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.

=cut
